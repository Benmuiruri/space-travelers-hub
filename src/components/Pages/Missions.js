/* eslint-disable consistent-return */
import { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import Table from 'react-bootstrap/Table';
import {
  getAllMissions,
  joinMission,
  leaveMission,
} from '../../redux/missions/missions';
import classes from './Missions.module.css';

const Missions = () => {
  // @ts-ignore
  const missions = useSelector((state) => state.missions);
  const dispatch = useDispatch();
  useEffect(() => {
    // @ts-ignore
    dispatch(getAllMissions());
  }, []);

  if (missions.length) {
    return (
      <section className={classes.missionsContainer}>
        <h2>Missions Page</h2>
        <Table striped bordered hover>
          <thead>
            <tr>
              <th>Mission Name</th>
              <th>Mission Description</th>
              <th>Mission Status</th>
              <th>Actions</th>
            </tr>
          </thead>
          <tbody>
            {missions.map((mission) => (
              <tr key={mission.id}>
                <td>{mission.name}</td>
                <td>{mission.description}</td>
                <td>
                  {!mission.joined && <span className={`${classes.status} ${classes['non-member']}`}> Not a member </span>}
                  {mission.joined && <span className={`${classes.status} ${classes['active-member']}`}> Active member </span>}
                </td>
                <td>
                  {!mission.joined && (
                    <button
                      type="button"
                      className={`${classes.missionBtn} ${classes['join-button']}`}
                      onClick={() => {
                        dispatch(joinMission(mission.id));
                      }}
                    >
                      Join Mission
                    </button>
                  )}
                  {mission.joined && (
                    <button
                      type="button"
                      className={`${classes.missionBtn} ${classes['leave-button']}`}
                      onClick={() => {
                        dispatch(leaveMission(mission.id));
                      }}
                    >
                      Leave Mission
                    </button>
                  )}
                </td>
              </tr>
            ))}
            ;
          </tbody>
        </Table>
      </section>
    );
  }
  return (
    <>
      <p>There are no missions </p>
    </>
  );
};
export default Missions;
