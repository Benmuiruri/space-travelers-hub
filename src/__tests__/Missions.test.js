import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../redux/configureStore';
import Missions from '../components/Pages/Missions';

// @ts-ignore
global.fetch = () => Promise.resolve({
  json: () => Promise.resolve([{
    id: 1,
    name: 'Apollo Mission',
    description: 'The famous Apolo mission',
    joined: false,
  }]),
});
const TestMissions = () => (
  <Provider store={store}>
    <Missions />
  </Provider>
);

describe('Testing Mission Component', () => {
  it('gets header in rendered Test Missions Component', () => {
    render(<TestMissions />);
    const element = screen.getByRole('heading', { name: 'Missions Page' });
    expect(element).toBeVisible();
  });
});

describe('Test presence of Missions component', () => {
  test('renders the Missions component correctly', () => {
    const missionComponent = renderer.create(<TestMissions />).toJSON();
    expect(missionComponent).toMatchSnapshot();
  });
});

describe('Test Join mission Button', () => {
  test('Join Button changes after it is clicked', () => {
    render(<TestMissions />);
    fireEvent.click(screen.getByRole('button', { name: 'Join Mission' }));
    const spanElement = screen.getByRole('cell', { name: /Active member/i });
    expect(spanElement).toBeVisible();
  });
});

describe('Test Leave mission Button', () => {
  test('Leave Button changes after it is clicked', () => {
    render(<TestMissions />);
    fireEvent.click(screen.getByRole('button', { name: 'Leave Mission' }));
    const spanElement = screen.getByRole('cell', { name: /Not a member/i });
    expect(spanElement).toBeVisible();
  });
});
