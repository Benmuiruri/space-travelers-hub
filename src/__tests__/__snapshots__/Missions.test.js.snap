// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test presence of Missions component renders the Missions component correctly 1`] = `
<section
  className="missionsContainer"
>
  <h2
    className="header"
  >
    Missions Page
  </h2>
  <table
    className="table table-striped table-bordered table-hover"
  >
    <thead>
      <tr>
        <th>
          Mission Name
        </th>
        <th>
          Mission Description
        </th>
        <th>
          Mission Status
        </th>
        <th>
          Actions
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td />
        <td>
          The famous Apolo mission
        </td>
        <td>
          <span
            className="status non-member"
          >
             
            Not a member
             
          </span>
        </td>
        <td>
          <button
            className="missionBtn join-button"
            onClick={[Function]}
            type="button"
          >
            Join Mission
          </button>
        </td>
      </tr>
      ;
    </tbody>
  </table>
</section>
`;
